Flaky checkstyle plugin run

### Apache Iceberg version

main (development)

### Query engine

None

### Please describe the bug 🐞

Too many times, I am facing one flaky error (also other people are facing it too). Unable to know the reason. 

* What went wrong:
Execution failed for task ':iceberg-spark:iceberg-spark-3.3_2.12:checkstyleMain'.
> A failure occurred while executing org.gradle.api.plugins.quality.internal.CheckstyleAction
   > An unexpected error occurred configuring and executing Checkstyle.
      > java.lang.Error: Error was thrown while processing /home/runner/work/iceberg/iceberg/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java
Failure details:
https://github.com/apache/iceberg/actions/workflows/java-ci.yml

https://github.com/apache/iceberg/actions/runs/4309428443
https://github.com/apache/iceberg/actions/runs/4310045826
https://github.com/apache/iceberg/actions/runs/4306392493/
and many more

```
Caused by: java.lang.OutOfMemoryError: GC overhead limit exceeded
	at org.antlr.v4.runtime.atn.ATNSimulator.getCachedContext(ATNSimulator.java:101)
	at org.antlr.v4.runtime.atn.ATNConfigSet.optimizeConfigs(ATNConfigSet.java:217)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.addDFAState(ParserATNSimulator.java:2124)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.addDFAEdge(ParserATNSimulator.java:2078)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.computeTargetState(ParserATNSimulator.java:619)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.execATN(ParserATNSimulator.java:454)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.adaptivePredict(ParserATNSimulator.java:393)
	at com.puppycrawl.tools.checkstyle.grammar.java.JavaLanguageParser.classOrInterfaceType(JavaLanguageParser.java:3838)
```

Didn't find any issue related to GC in https://github.com/checkstyle/checkstyle

Planning to increase the gradle memory limits. Suggestions are welcome.

--- a/build.gradle
+++ b/build.gradle
@@ -32,6 +32,10 @@ buildscript {
     // in an old version of the errorprone, which doesn't work w/ Gradle 8, so bump errorpone as
     // well.
     classpath "net.ltgt.gradle:gradle-errorprone-plugin:3.0.1"
+    // com.palantir.baseline:gradle-baseline-java:4.42.0 (the last version supporting Java 8) pulls
+    // in an old version of the checkstyle(9.1), which has this OutOfMemory bug https://github.com/checkstyle/checkstyle/issues/10934
+    // So, bump checkstyle to the latest java 8 supported version(9.3) which contains the fix.
+    classpath "com.puppycrawl.tools:checkstyle:9.3"
 
     classpath 'com.diffplug.spotless:spotless-plugin-gradle:6.13.0'
     classpath 'gradle.plugin.org.inferred:gradle-processors:3.7.0'

